### PR TITLE
Android targets create final zip artifacts

### PR DESCRIPTION
**Pull request by dnfield**
This resolves the Android part of https://github.com/flutter/flutter/issues/42140

This seems like a good stopping point to make the patch not become huge.  iOS should also be a fairly small patch.  The host targets might be a bit more involved.

The goal is to land this, which would then enable updating the recipes to be much simpler (and make sure that breakages around what files are included are caught earlier and fixed).

The javadoc targets error out right now, filed https://github.com/flutter/flutter/issues/42400 to track.

This also doesn't create the proposed file to tell CI where to find these artifacts. I'm going to need to think a bit about how to make that work when different targets are involved.

### CODE REVIEW

File: build/zip_bundle.gni [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import("//flutter/common/config.gni")
+
+if (flutter_runtime_mode == "jit_release") {
+  android_zip_archive_dir = "android-$target_cpu-jit-release"
+} else {
+  android_zip_archive_dir = "android-$target_cpu"
+}
+
+template("zip_bundle") {
+  action(target_name) {
```
Comment (chinmaygarde) [inline build/zip_bundle.gni]: Please document template args in one or both ways:

* An `assert(defined(invoker.foo), "Foo must be defined")` declaration before doing anything so the required template arguments are asserted to exit.
* A comment block above the template that says what the template does and documents the required (and optional) invoker arguments.

File: build/zip.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+      zip_file.write(path, archive_name)
+  zip_file.close()
+
+
+if __name__ == '__main__':
+  parser = argparse.ArgumentParser(
+      description='This script creates zip files.')
+  parser.add_argument('-o', action='store',
+      help='The name of the output zip file.')
+  parser.add_argument('-i', nargs=2, action='append',
```
Comment (chinmaygarde) [inline build/zip.py]: Please specify the destination variable in argparser arguments via `dest='source_path_archive_path_pair'`, etc.. That way, the function above is easier to parse. I wasn't sure what `args.i` meant.

File: build/zip.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+# found in the LICENSE file.
+
+import argparse
+import zipfile
+import os
+import sys
+
+
+def _zip_dir(path, zip_file, prefix):
+  path = path.rstrip('/\\')
```
Comment (chinmaygarde) [inline build/zip.py]: Not sure why this is necessary.

File: build/zip.py (hunk 2 above)
Comment (dnfield) [reply build/zip.py]: Done

File: build/zip_bundle.gni (hunk 1 above)
Comment (dnfield) [reply build/zip_bundle.gni]: Done

File: build/zip.py (hunk 3 above)
Comment (dnfield) [reply build/zip.py]: So that I can `root.replace` using it and not end up with extra slashes.

Comment (dnfield) [reply build/zip.py]: (in case someone gives a directory as the argument and includes the extra slash)

**Pull request merged**